### PR TITLE
Refactor Level Zero Adapter to use new logger

### PR DESCRIPTION
**Pull request by lplewa**

### CODE REVIEW

File: source/adapters/level_zero/adapter.hpp [hunk 1]
```diff
@@ -10,10 +10,14 @@
...
 
 struct ur_adapter_handle_t_ {
   std::atomic<uint32_t> RefCount = 0;
   std::mutex Mutex;
+  logger::Logger logger = logger::create_logger("level_zero");
 };
 
 extern ur_adapter_handle_t_ Adapter;
+
+
```
Comment (lukaszstolarczuk) [inline source/adapters/level_zero/adapter.hpp:23]: redundant extra empty lines

File: source/adapters/level_zero/command_buffer.cpp [hunk 2]
```diff
@@ -180,16 +181,17 @@ ur_result_t calculateKernelWorkDimensions(
...
           return UR_RESULT_ERROR_INVALID_WORK_GROUP_SIZE;
         }
         WG[I] = GroupSize[I];
       }
-      urPrint(
-          "urCommandBufferAppendKernelLaunchExp: using computed WG size = {%d, "
-          "%d, %d}\n",
-          WG[0], WG[1], WG[2]);
+      logger::debug("urCommandBufferAppendKernelLaunchExp: using computed WG "
+                    "size = {{{}, "
```
Comment (lukaszstolarczuk) [inline source/adapters/level_zero/command_buffer.cpp:192]: it seems like you'd have an extra pair of `{` - is this expected?

actually, general question - have you tested all these works? 😄

File: source/adapters/level_zero/command_buffer.cpp [hunk 3]
```diff
@@ -217,30 +219,33 @@ ur_result_t calculateKernelWorkDimensions(
     break;
 
   default:
-    urPrint("urCommandBufferAppendKernelLaunchExp: unsupported work_dim\n");
+    logger::error("urCommandBufferAppendKernelLaunchExp: unsupported work_dim");
```
Comment (lukaszstolarczuk) [inline source/adapters/level_zero/command_buffer.cpp:222]: here you removed the `\n`, above you left it as it was. My guess is you want to remove all these `\n` (I spotted at least several other).

File: source/adapters/level_zero/command_buffer.cpp (hunk 2 above)
Comment (lplewa) [reply source/adapters/level_zero/command_buffer.cpp:192]: Double {{ means a single '{'

File: source/adapters/level_zero/command_buffer.cpp (hunk 3 above)
Comment (lplewa) [reply source/adapters/level_zero/command_buffer.cpp:222]: \n is added by logger so all \n are redundant

File: source/adapters/level_zero/adapter.hpp (hunk 1 above)
Comment (DamianDuy) [reply source/adapters/level_zero/adapter.hpp:23]: you can use `make cppformat` command to remove stuff like that automatically

File: source/adapters/level_zero/adapter.cpp [hunk 4]
```diff
@@ -184,7 +184,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urAdapterGetLastError(
   std::ignore = Adapter;
   std::ignore = Message;
   std::ignore = Error;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
+  logger::error("[UR][L0] {} function not implemented!", __FUNCTION__);
```
Comment (PatKamin) [inline source/adapters/level_zero/adapter.cpp:187]: As the name of the logger is `level_zero`, the beginning of this logger output will look like that, I think: `<level_zero>[ERROR][UR][L0]`. No need for double level zero info, I suppose.

File: source/adapters/level_zero/command_buffer.cpp [hunk 5]
```diff
@@ -180,16 +181,17 @@ ur_result_t calculateKernelWorkDimensions(
...
         }
         WG[I] = GroupSize[I];
       }
-      urPrint(
-          "urCommandBufferAppendKernelLaunchExp: using computed WG size = {%d, "
-          "%d, %d}\n",
-          WG[0], WG[1], WG[2]);
+      logger::debug("urCommandBufferAppendKernelLaunchExp: using computed WG "
+                    "size = {{{}, "
+                    "{}, {}}}\n",
```
Comment (PatKamin) [inline source/adapters/level_zero/command_buffer.cpp:193]: Is the double newline (one explicit and one aded by the logger) intended?

File: source/adapters/level_zero/context.cpp [hunk 6]
```diff
@@ -174,7 +175,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urContextSetExtendedDeleter(
   std::ignore = Context;
   std::ignore = Deleter;
   std::ignore = UserData;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
+  logger::error("[UR][L0] {} function not implemented!", __FUNCTION__);
```
Comment (PatKamin) [inline source/adapters/level_zero/context.cpp:178]: Please, test these `__FUNCTION__` prints with different compilers, perhaps create a test case for such a print to verify if this works with compilers tested in UR CI.

File: source/adapters/level_zero/device.cpp [hunk 7]
```diff
@@ -980,7 +981,7 @@ ur_result_t ur_device_handle_t_::initialize(int SubSubDeviceOrdinal,
   if (numQueueGroups == 0) {
     return UR_RESULT_ERROR_UNKNOWN;
   }
-  urPrint("NOTE: Number of queue groups = %d\n", numQueueGroups);
+  logger::error("NOTE: Number of queue groups = {}\n", numQueueGroups);
```
Comment (PatKamin) [inline source/adapters/level_zero/device.cpp:984]: Extra newline, is it needed?

File: source/adapters/level_zero/event.cpp [hunk 8]
```diff
@@ -742,7 +741,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urEventSetCallback(
   std::ignore = ExecStatus;
   std::ignore = Notify;
   std::ignore = UserData;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
+  logger::error("[UR][L0] {} function not implemented!\n", __FUNCTION__);
```
Comment (PatKamin) [inline source/adapters/level_zero/event.cpp:744]: Remove an extra newline.

File: source/adapters/level_zero/platform.cpp [hunk 9]
```diff
@@ -69,7 +70,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urPlatformGet(
   }
 
   if (ZeResult != ZE_RESULT_SUCCESS) {
-    urPrint("zeInit: Level Zero initialization failure\n");
+    logger::debug("zeInit: Level Zero initialization failure");
```
Comment (PatKamin) [inline source/adapters/level_zero/platform.cpp:73]: Sounds like at least a warning message, not debug.

File: source/adapters/level_zero/sampler.cpp [hunk 10]
```diff
@@ -75,9 +76,9 @@ UR_APIEXPORT ur_result_t UR_APICALL urSamplerCreate(
       ZeSamplerDesc.addressMode = ZE_SAMPLER_ADDRESS_MODE_MIRROR;
       break;
     default:
-      urPrint("urSamplerCreate: unsupported "
-              "UR_SAMPLER_PROPERTIES_ADDRESSING_MODEE "
-              "value\n");
+      logger ::error("urSamplerCreate: unsupported "
```
Comment (PatKamin) [inline source/adapters/level_zero/sampler.cpp:79]: nit: Remove a whitespace after the namespace name.

Comment (lplewa) [reply source/adapters/level_zero/sampler.cpp]: Dpne

File: source/adapters/level_zero/platform.cpp (hunk 9 above)
Comment (lplewa) [reply source/adapters/level_zero/platform.cpp]: Done

File: source/adapters/level_zero/event.cpp (hunk 8 above)
Comment (lplewa) [reply source/adapters/level_zero/event.cpp:744]: Done

File: source/adapters/level_zero/device.cpp (hunk 7 above)
Comment (lplewa) [reply source/adapters/level_zero/device.cpp]: Done

File: source/adapters/level_zero/context.cpp (hunk 6 above)
Comment (lplewa) [reply source/adapters/level_zero/context.cpp]: Why do you think that it will not work?

File: source/adapters/level_zero/command_buffer.cpp (hunk 5 above)
Comment (lplewa) [reply source/adapters/level_zero/command_buffer.cpp]: Done

File: source/adapters/level_zero/adapter.cpp (hunk 4 above)
Comment (lplewa) [reply source/adapters/level_zero/adapter.cpp]: Done

File: source/adapters/level_zero/command_buffer.cpp (hunk 2 above)
Comment (lukaszstolarczuk) [reply source/adapters/level_zero/command_buffer.cpp:192]: I see, thx. Nonetheless, can you paste a few example outputs from these logs? 😉

File: source/adapters/level_zero/event.cpp (hunk 8 above)
Comment (lukaszstolarczuk) [reply source/adapters/level_zero/event.cpp:744]: I believe `\n` is still here (I hope Patryk was talking about this, not something else)

Comment (PatKamin) [reply source/adapters/level_zero/event.cpp:744]: Yes, that is what I meant.

**Comment by PatKamin** [comment]
> Why do you think that it will not work? 

I'm not sure if it will print a kind of "pretty" name of the function. I mean not "heavily" mangled in such a way that it's uncertain what the name of the function is. Could you paste some example logs?

**Comment by lplewa** [comment]
> > Why do you think that it will not work?
> 
> I'm not sure if it will print a kind of "pretty" name of the function. I mean not "heavily" mangled in such a way that it's uncertain what the name of the function is. Could you paste some example logs?

Technically it is compiler specific, but i tested few compilers and it's allways not mangled 
I think this example will make better job as an example. 
image

File: source/adapters/level_zero/kernel.cpp [hunk 11]
```diff
@@ -482,8 +484,8 @@ UR_APIEXPORT ur_result_t UR_APICALL urKernelGetInfo(
       return UR_RESULT_ERROR_UNKNOWN;
     }
   default:
-    urPrint("Unsupported ParamName in urKernelGetInfo: ParamName=%d(0x%x)\n",
-            ParamName, ParamName);
+    logger::error("Unsupported ParamName in urKernelGetInfo: ParamName={}\n",
```
Comment (PatKamin) [inline source/adapters/level_zero/kernel.cpp:487]: Remove `\n`

File: source/adapters/level_zero/kernel.cpp [hunk 12]
```diff
@@ -541,8 +543,8 @@ UR_APIEXPORT ur_result_t UR_APICALL urKernelGetGroupInfo(
     return ReturnValue(uint32_t{Kernel->ZeKernelProperties->privateMemSize});
   }
   default: {
-    urPrint("Unknown ParamName in urKernelGetGroupInfo: ParamName=%d(0x%x)\n",
-            ParamName, ParamName);
+    logger::error("Unknown ParamName in urKernelGetGroupInfo: ParamName=%d\n",
```
Comment (PatKamin) [inline source/adapters/level_zero/kernel.cpp:546]: Remove `\n`

File: source/adapters/level_zero/platform.cpp [hunk 13]
```diff
@@ -48,7 +48,8 @@ UR_APIEXPORT ur_result_t UR_APICALL urPlatformGet(
   }
 
   if (getenv("SYCL_ENABLE_PCI") != nullptr) {
-    urPrint("WARNING: SYCL_ENABLE_PCI is deprecated and no longer needed.\n");
+    logger::warning(
+        "WARNING: SYCL_ENABLE_PCI is deprecated and no longer needed.");
```
Comment (PatKamin) [inline source/adapters/level_zero/platform.cpp:52]: Remove the `WARNING` word as the logger should provide this information.

File: source/adapters/level_zero/memory.cpp [hunk 14]
```diff
@@ -2020,7 +2022,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urMemImageGetInfo(
   std::ignore = PropSize;
   std::ignore = ImgInfo;
   std::ignore = PropSizeRet;
-  urPrint("[UR][L0] %s function not implemented!\n", __FUNCTION__);
+  logger::debug("{} function not implemented!", __FUNCTION__);
```
Comment (PatKamin) [inline source/adapters/level_zero/memory.cpp:2025]: Some of this kind of messages (`function not implemented`) are of the `debug` type whereas most are of the `error` type (in other files, too). Shouldn't they be unified and have the same log level across the level zero?

File: source/adapters/level_zero/event.cpp (hunk 8 above)
Comment (lplewa) [reply source/adapters/level_zero/event.cpp]: This time for sure - Done

File: source/adapters/level_zero/memory.cpp (hunk 14 above)
Comment (lplewa) [reply source/adapters/level_zero/memory.cpp]: Done

File: source/adapters/level_zero/platform.cpp (hunk 13 above)
Comment (lplewa) [reply source/adapters/level_zero/platform.cpp]: Done

File: source/adapters/level_zero/kernel.cpp (hunk 12 above)
Comment (lplewa) [reply source/adapters/level_zero/kernel.cpp]: Done

File: source/adapters/level_zero/kernel.cpp (hunk 11 above)
Comment (lplewa) [reply source/adapters/level_zero/kernel.cpp]: Done

File: source/adapters/level_zero/command_buffer.cpp (hunk 2 above)
Comment (lplewa) [reply source/adapters/level_zero/command_buffer.cpp:194]:
```
<LOADER>[DEBUG]: Logger loader initialized successfully!
Platform initialized.
<LEVEL_ZERO>[DEBUG]: ZE ---> zeInit(ZE_INIT_FLAG_GPU_ONLY)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGet(&ZeDriverCount, nullptr)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGet(&ZeDriverCount, ZeDrivers.data())
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGetProperties(ZeDriver, &ZeDriverProperties)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGetApiVersion(ZeDriver, &ZeApiVersion)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGetExtensionProperties(ZeDriver, &Count, nullptr)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGetExtensionProperties(ZeDriver, &Count, ZeExtensions.data())
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGetExtensionFunctionAddress(DriverHandle, "zexDriverImportExternalPointer", reinterpret_cast<void **>(&zexDriverImportExternalPointer))
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDriverGetExtensionFunctionAddress(DriverHandle, "zexDriverReleaseImportedPointer", reinterpret_cast<void **>(&zexDriverReleaseImportedPointer))
API version: 0.7
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGet(ZeDriver, &ZeDeviceCount, nullptr)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGet(ZeDriver, &ZeDeviceCount, ZeDevices.data())
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGetProperties(ZeDevice, &Properties)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGetCommandQueueGroupProperties(ZeDevice, &numQueueGroups, nullptr)
<LEVEL_ZERO>[INFO]: Number of queue groups = 1
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGetCommandQueueGroupProperties(ZeDevice, &numQueueGroups, QueueGroupProperties.data())
<LEVEL_ZERO>[INFO]: main blitter/copy engine is not available
<LEVEL_ZERO>[INFO]: link blitter/copy engines are not available
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGetSubDevices(Device->ZeDevice, &SubDevicesCount, nullptr)
<LEVEL_ZERO>[DEBUG]: ZE ---> zeDeviceGetSubDevices(Device->ZeDevice, &SubDevicesCount, ZeSubd
```

**Review by PatKamin — APPROVED** [review]

File: source/adapters/level_zero/event.cpp [hunk 15]
```diff
@@ -15,18 +15,17 @@
...
 #include "common.hpp"
 #include "event.hpp"
+#include "logger/ur_logger.hpp"
 #include "ur_level_zero.hpp"
 
 void printZeEventList(const _ur_ze_event_list_t &UrZeEventList) {
-  if (UrL0Debug & UR_L0_DEBUG_BASIC) {
-    urPrint("  NumEventsInWaitList %d:", UrZeEventList.Length);
+  std::stringstream ss;
+  ss << "  NumEventsInWaitList" << UrZeEventList.Length << ":";
```
Comment (lukaszstolarczuk) [inline source/adapters/level_zero/event.cpp:23]: you're missing a space after `NumEventsInWaitList`

File: source/adapters/level_zero/event.cpp [hunk 16]
```diff
@@ -15,18 +15,17 @@
...
 
-    for (uint32_t I = 0; I < UrZeEventList.Length; I++) {
-      urPrint(" %#llx", ur_cast<std::uintptr_t>(UrZeEventList.ZeEventList[I]));
-    }
-
-    urPrint("\n");
```
Comment (lukaszstolarczuk) [inline source/adapters/level_zero/event.cpp:28]: previously you had a new line after each entry, now you have it all printed at once (with spaces between) - is this expected?

Comment (lplewa) [reply source/adapters/level_zero/event.cpp:28]: There was only one new line

Comment (lukaszstolarczuk) [reply source/adapters/level_zero/event.cpp:28]: haha, mea culpa, I'm blind, apparently 😉  // the second `}` (from `if`) got me, I think

**Review by isaacault — APPROVED** [review]
`source/adapters/level_zero/image.cpp` LGTM

**Review by przemektmalon — APPROVED** [review]
`image.cpp` changes LGTM

**Comment by jandres742** [comment]
@lplewa : do we have intel/llvm PR testing this? I want to make sure there are no sycl tests relying on those strings. If so, we would need to have the intel/llvm PR to modify also the tests.

**Comment by kswiecicki** [comment]
@lplewa the SYCL testing process for adapter changes is available here: https://github.com/oneapi-src/unified-runtime/blob/main/scripts/core/CONTRIB.rst#adapter-change-process

**Comment by fabiomestre** [comment]
I have updated the target branch of this PR from the `adapters` branch to the `main` branch.
Development in UR is moving back to `main`. The `adapters` branch will soon be deleted.

File: source/adapters/cuda/common.cpp [hunk 17]
```diff
@@ -37,29 +38,34 @@ ur_result_t mapErrorUR(CUresult Result) {
...
+  static bool abort = std::getenv("PI_CUDA_ABORT") != nullptr ||
+                      std::getenv("UR_CUDA_ABORT") != nullptr;
+
   if (Result == CUDA_SUCCESS || Result == CUDA_ERROR_DEINITIALIZED) {
     return;
   }
 
-  if (std::getenv("SYCL_PI_SUPPRESS_ERROR_MESSAGE") == nullptr &&
-      std::getenv("UR_SUPPRESS_ERROR_MESSAGE") == nullptr) {
+  if (suppress_error) {
```
Comment (JackAKirk) [inline source/adapters/cuda/common.cpp:52]: It looks like if `suppress_error` is true it doesn't suppress the error, right?

Shouldn't it be then named differently or defined differently/used differently?

**Comment by lplewa** [comment]
PR was updated, with a legacy compatibility mode, so we don't have to update all tests in this repository.

image

File: source/common/logger/ur_sinks.hpp [hunk 18]
```diff
@@ -50,6 +45,14 @@ class Sink {
         flush_level = logger::Level::ERR;
     }
 
+    virtual void print(logger::Level level, std::string msg) {
```
[the remark this answers is not in the document]
Comment (lplewa) [reply source/common/logger/ur_sinks.hpp:48]: No - we need c_str() for "legacy" compatibility

**Review by PatKamin — APPROVED** [review]

**Comment by JackAKirk** [comment]
This no longer touches any cuda/hip code. So don't think my review is required. I am a bit confused about why this error messaging system is different in l0 compared to other backends. And I also wonder whether it is being wired up to dpc++ properly in order to output sycl exceptions, as per the SYCL spec....

**Comment by lplewa** [comment]
> This no longer touches any cuda/hip code. So don't think my review is required. I am a bit confused about why this error messaging system is different in l0 compared to other backends. And I also wonder whether it is being wired up to dpc++ properly in order to output sycl exceptions, as per the SYCL spec....

I decided to update adapters one by one to simplify review. Cuda will be updated to new logger in next PR

**Comment by JackAKirk** [comment]
> > This no longer touches any cuda/hip code. So don't think my review is required. I am a bit confused about why this error messaging system is different in l0 compared to other backends. And I also wonder whether it is being wired up to dpc++ properly in order to output sycl exceptions, as per the SYCL spec....
> 
> I decided to update adapters one by one to simplify review. Cuda will be updated to new logger in next PR

This
```
 virtual void print([[maybe_unused]] logger::Level level, std::string msg) {
    fprintf(stderr, "%s", msg.c_str());
  }
```

is just directly printing errors/warnings. Which I think means that when used by the sycl programming model it isn't compliant with the sycl specification where errors are passed via `sycl::exception`: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_error_handling

It seems to generally be another implementation to report plugin specific errors. I'm confused because I thought there was already an agreed mechanism for this that is already being used by DPC++ to correctly return `sycl::exception`, I thought it had been agreed that the various mish-mash of error reporting was going to move to this system. see e.g. https://github.com/oneapi-src/unified-runtime/issues/500 for some discussion. Is there a good reason for this duplication?

**Comment by ldrumm** [comment]
> No description provided

I'm fine with the changes, but have no idea of the rationale. The commit message needs to provide the future reader with some context, please

**Review by ldrumm — CHANGES_REQUESTED** [review]
Add a commit message with some context

**Comment by pbalcer** [comment]
> [](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_error_handling)
> 
> It seems to generally be another implementation to report plugin specific errors. I'm confused because I thought there was already an agreed mechanism for this that is already being used by DPC++ to correctly return `sycl::exception`, I thought it had been agreed that the various mish-mash of error reporting was going to move to this system. see e.g. #500 for some discussion. Is there a good reason for this duplication?

Right now this is just trying to consolidate how the loader and the adapters print messages. So replacing all the random `printf`/`fprintf` calls with one path that's a) simpler to use, and b) is controlled using the same [mechanism](https://oneapi-src.github.io/unified-runtime/core/INTRO.html#logging) for all components, resulting in hopefully better user experience. The logger is not specifically about error handling.

Would it make sense for there to be some mechanism that would take the latest logged error and return it in `urAdapterGetLastError`? Probably, but that's orthogonal to this patch.

**Comment by JackAKirk** [comment]
> > It seems to generally be another implementation to report plugin specific errors. I'm confused because I thought there was already an agreed mechanism for this that is already being used by DPC++ to correctly return `sycl::exception`, I thought it had been agreed that the various mish-mash of error reporting was going to move to this system. see e.g. #500 for some discussion. Is there a good reason for this duplication?
> 
> Right now this is just trying to consolidate how the loader and the adapters print messages. So replacing all the random `printf`/`fprintf` calls with one path that's a) simpler to use, and b) is controlled using the same [mechanism](https://oneapi-src.github.io/unified-runtime/core/INTRO.html#logging) for all components, hopefully resulting in better user experience. The logger is not specifically about error handling.
> 
> Would it make sense for there to be some mechanism that would take the latest logged error and return it in `urAdapterGetLastError`? Probably, but that's orthogonal to this patch.

Thanks for the explanation.

**Comment by lplewa** [comment]
> > > This no longer touches any cuda/hip code. So don't think my review is required. I am a bit confused about why this error messaging system is different in l0 compared to other backends. And I also wonder whether it is being wired up to dpc++ properly in order to output sycl exceptions, as per the SYCL spec....
> > 
> > 
> > I decided to update adapters one by one to simplify review. Cuda will be updated to new logger in next PR
> 
> This
> 
> ```
>  virtual void print([[maybe_unused]] logger::Level level, std::string msg) {
>     fprintf(stderr, "%s", msg.c_str());
>   }
> ```
> 
> is just directly printing errors/warnings. Which I think means that when used by the sycl programming model it isn't compliant with the sycl specification where errors are passed via `sycl::exception`: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_error_handling

This is Legacy compatibility mode for Level zero - it is needed as sycl e2e tests parses those logs. Those tests should be refactored to use new log API but for now it is easier to provide "compatibility mode" and then refactor sycl tests, instead doing this change at once in both repositories.

**Comment by lplewa** [comment]
I had to rebase this commit to older one, as unified-runtime/main currently do not compile with sycl
Here you can find build status of this PR with Sycl tests
https://github.com/intel/llvm/pull/12580

**Comment by lplewa** [comment]
I spited this PR to two PR
First part is here:
https://github.com/oneapi-src/unified-runtime/pull/1327
and when it will be merged i will rebase this PR to include only changes in levelzero adapter. This will allow me to create PR with changes in other adapters without dependencies, to this PR

**Comment by lplewa** [comment]
I created an octopus merge b45470f9e1a5c4bacbc210d68e900858b82fae25 of all logs changes (#1359; #1347; #1345; #1342; #1032) and tested integration with [intel/llvm repository](https://github.com/intel/llvm/pull/12580/files)

**Pull request merged**